src/main/java/org/jetbrains/nativecerts/mac/SecurityFrameworkUtil.java

```diff
@@ -43,6 +63,19 @@ public static List<X509Certificate> getTrustedRoots(SecurityFramework.SecTrustSe
         return result;
     }
 
+    public static List<X509Certificate> SecItemCopyMatching(
+            CoreFoundation.CFDictionaryRef query,
+            Predicate<SecurityFramework.SecCertificateRef> predicate
+    ) {
+        CFArrayRefByReference returnedCertArray = new CFArrayRefByReference();
+        SecurityFramework.OSStatus rc = SecurityFramework.INSTANCE.SecItemCopyMatching(query, returnedCertArray);
+        if (!SecurityFramework.OSStatus.errSecSuccess.equals(rc)) {
```

Comment:
currently getting `One or more parameters passed to a function were not valid.`

API signature: https://developer.apple.com/documentation/security/secitemcopymatching(_:_:)?language=objc

Working native code:

```
    CFTypeRef searchKeys[] = { kSecClass, kSecMatchLimit, kSecReturnRef, kSecMatchSearchList };
    CFTypeRef searchValues[] = { kSecClassCertificate, kSecMatchLimitAll, kCFBooleanTrue, keychainList };
    search = CFDictionaryCreate(kCFAllocatorDefault,
                              searchKeys, searchValues, 4, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
    if (search == NULL) {
        goto errOut;
    }

    if( SecItemCopyMatching( search, (CFTypeRef *)&currAnchors ) == errSecSuccess ){
```

Reply:
Hi @shalupov

I see you've done most of the work on this library.

Any chance you have a pointer to where I'm going wrong here.

Thanks!

Reply:
keychainList should not be a constant here. https://github.com/openjdk/jdk/blob/59ceebc4acf2c39c86c45b1aa6746bf619417cea/src/java.base/macosx/native/libosxsecurity/KeystoreImpl.m#L558 KeystoreImpl gets keychain list from some predefined list, but I'm worrying thar it's deprecated.

(some random forum message) https://developer.apple.com/forums/thread/739774

Reply:
oh that was silly of me...

I don't think its needed anyway, I didn't implement that in node.js:

https://github.com/nodejs/node/pull/56599

Reply:
removing it doesn't make a difference, still get the useless `One or more parameters passed to a function were not valid`